File: scripts/optimize-videos.mjs
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { execFile } from 'child_process';
import { promisify } from 'util';

const execFileAsync = promisify(execFile);

// Dynamically import ffmpeg-static path
let ffmpegPath;
try {
  const ffmpegStatic = await import('ffmpeg-static');
  ffmpegPath = ffmpegStatic.default || ffmpegStatic;
} catch (err) {
  console.error('\n❌ ERROR: "ffmpeg-static" module is missing. Please run:\n   npm install ffmpeg-static\n');
  process.exit(1);
}

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const ROOT_DIR = path.resolve(__dirname, '..');

const IGNORE_DIRS = new Set([
  'node_modules',
  '.git',
  '.next',
  'dist',
  'build',
  'out',
  '.vscode',
  '.gemini',
  'brain'
]);

const VIDEO_GIF_EXTENSIONS = new Set(['.mp4', '.mov', '.avi', '.mkv', '.webm', '.m4v', '.gif']);
const CODE_EXTENSIONS = new Set(['.html', '.css', '.js', '.jsx', '.ts', '.tsx', '.json', '.md']);

/**
 * Format bytes into human-readable string (KB/MB)
 */
function formatBytes(bytes) {
  if (bytes === 0) return '0 B';
  const k = 1024;
  const sizes = ['B', 'KB', 'MB', 'GB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i];
}

/**
 * Compress and convert video/GIF to WebM (VP9) or MP4 (H.264)
 */
async function compressMedia(inputPath, outputPath, isGif = false, format = 'mp4') {
  let args = [];

  if (isGif) {
    // Convert GIF to MP4 / WebM with silent audio & yuv420p color space
    if (format === 'webm') {
      args = [
        '-y',
        '-i', inputPath,
        '-c:v', 'libvpx-vp9',
        '-b:v', '0',
        '-crf', '32',
        '-pix_fmt', 'yuva420p',
        '-an',
        outputPath
      ];
    } else {
      args = [
        '-y',
        '-i', inputPath,
        '-movflags', 'faststart',
        '-pix_fmt', 'yuv420p',
        '-vf', 'scale=trunc(iw/2)*2:trunc(ih/2)*2',
        '-c:v', 'libx264',
        '-crf', '24',
        '-preset', 'medium',
        '-an',
        outputPath
      ];
    }
  } else {
    // Video Optimization (Preserves audio track, H.264 faststart or WebM VP9)
    if (format === 'webm') {
      args = [
        '-y',
        '-i', inputPath,
        '-c:v', 'libvpx-vp9',
        '-crf', '30',
        '-b:v', '0',
        '-c:a', 'libopus',
        '-b:a', '128k',
        outputPath
      ];
    } else {
      args = [
        '-y',
        '-i', inputPath,
        '-c:v', 'libx264',
        '-crf', '23',
        '-preset', 'medium',
        '-movflags', '+faststart',
        '-c:a', 'aac',
        '-b:a', '128k',
        '-vf', 'scale=trunc(iw/2)*2:trunc(ih/2)*2',
        outputPath
      ];
    }
  }

  await execFileAsync(ffmpegPath, args);
}

/**
 * Scan directory recursively
 */
async function scanDirectory(dir, mediaFiles = [], codeFiles = []) {
  const entries = await fs.promises.readdir(dir, { withFileTypes: true });

  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);

    if (entry.isDirectory()) {
      if (!IGNORE_DIRS.has(entry.name)) {
        await scanDirectory(fullPath, mediaFiles, codeFiles);
      }
    } else if (entry.isFile()) {
      const ext = path.extname(entry.name).toLowerCase();
      if (VIDEO_GIF_EXTENSIONS.has(ext)) {
        mediaFiles.push(fullPath);
      } else if (CODE_EXTENSIONS.has(ext)) {
        codeFiles.push(fullPath);
      }
    }
  }
  return { mediaFiles, codeFiles };
}

/**
 * Main execution script
 */
async function run() {
  console.log('🎥 Starting Video & GIF Optimization Automation...\n');
  console.log(`Using FFmpeg binary at: ${ffmpegPath}\n`);

  const { mediaFiles, codeFiles } = await scanDirectory(ROOT_DIR);

  if (mediaFiles.length === 0) {
    console.log('✨ No video or GIF files found to optimize.');
    return;
  }

  console.log(`📹 Found ${mediaFiles.length} media file(s) to optimize.`);
  console.log(`📝 Found ${codeFiles.length} code file(s) to check for references.\n`);

  const report = [];
  const conversionMap = new Map();
  let totalOriginalSize = 0;
  let totalOptimizedSize = 0;

  // Phase 1: Convert & Compress Media
  for (const oldPath of mediaFiles) {
    const parsed = path.parse(oldPath);
    const isGif = parsed.ext.toLowerCase() === '.gif';

    // Target extension: .mp4 (or .webm based on requirement)
    const targetExt = isGif ? '.mp4' : parsed.ext.toLowerCase();
    const tempOutputPath = path.join(parsed.dir, `${parsed.name}_optimized${targetExt}`);
    const finalOutputPath = path.join(parsed.dir, `${parsed.name}${targetExt}`);

    const relOld = path.relative(ROOT_DIR, oldPath);
    console.log(`Optimizing: ${relOld}...`);

    try {
      const originalStat = await fs.promises.stat(oldPath);
      const originalSize = originalStat.size;
      totalOriginalSize += originalSize;

      // Perform FFmpeg Compression
      await compressMedia(oldPath, tempOutputPath, isGif, targetExt === '.webm' ? 'webm' : 'mp4');

      const optimizedStat = await fs.promises.stat(tempOutputPath);
      const optimizedSize = optimizedStat.size;

      // Replace target file if size improved or format converted
      if (oldPath !== finalOutputPath) {
        // If extension changed (e.g. .gif -> .mp4 or .mov -> .mp4)
        await fs.promises.rename(tempOutputPath, finalOutputPath);
      } else {
        // Overwrite original with optimized version
        await fs.promises.rename(tempOutputPath, oldPath);
      }

      totalOptimizedSize += optimizedSize;

      const reductionPercent = (((originalSize - optimizedSize) / originalSize) * 100).toFixed(2);
      const oldFilename = parsed.base;
      const newFilename = path.parse(finalOutputPath).base;

      conversionMap.set(oldFilename, newFilename);

      report.push({
        status: 'SUCCESS',
        originalName: oldFilename,
        optimizedName: newFilename,
        origSize: originalSize,
        optSize: optimizedSize,
        reduction: `${reductionPercent}%`,
        relPath: relOld
      });

      console.log(`  └─ ✅ Success: ${formatBytes(originalSize)} ➔ ${formatBytes(optimizedSize)} (${reductionPercent}% reduction)`);

    } catch (err) {
      console.error(`  └─ ❌ Failed to optimize ${relOld}:`, err.message);
      if (fs.existsSync(tempOutputPath)) {
        await fs.promises.unlink(tempOutputPath).catch(() => {});
      }
      report.push({
        status: 'FAILED',
        originalName: parsed.base,
        optimizedName: '-',
        origSize: 0,
        optSize: 0,
        reduction: '0%',
        relPath: relOld,
        error: err.message
      });
    }
  }

  // Phase 2: Update Code References
  console.log('\n🔄 Updating media references in source code files...');
  let updatedCodeFilesCount = 0;

  for (const codePath of codeFiles) {
    try {
      let content = await fs.promises.readFile(codePath, 'utf8');
      let modified = false;

      for (const [oldName, newName] of conversionMap.entries()) {
        if (oldName !== newName && content.includes(oldName)) {
          const regex = new RegExp(oldName.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'g');
          content = content.replace(regex, newName);
          modified = true;
        }
      }

      if (modified) {
        await fs.promises.writeFile(codePath, content, 'utf8');
        updatedCodeFilesCount++;
        console.log(`  └─ Updated: ${path.relative(ROOT_DIR, codePath)}`);
      }
    } catch (err) {
      console.error(`  └─ ❌ Could not update ${codePath}:`, err.message);
    }
  }

  // Phase 3: Cleanup old original files if extension was changed
  console.log('\n🧹 Cleaning up old original files if extension changed...');
  let deletedCount = 0;
  for (const [oldName, newName] of conversionMap.entries()) {
    if (oldName !== newName) {
      for (const oldPath of mediaFiles) {
        if (path.basename(oldPath) === oldName && fs.existsSync(oldPath)) {
          await fs.promises.unlink(oldPath).catch(() => {});
          deletedCount++;
        }
      }
    }
  }

  // Detailed Summary Log
  console.log('\n======================================================');
  console.log('📊 MEDIA OPTIMIZATION SUMMARY REPORT');
  console.log('======================================================');
  console.table(
    report.map((r) => ({
      'Original File': r.originalName,
      'Optimized File': r.optimizedName,
      'Original Size': formatBytes(r.origSize),
      'Optimized Size': formatBytes(r.optSize),
      'Reduction': r.reduction,
      'Status': r.status
    }))
  );

  const totalSaved = totalOriginalSize - totalOptimizedSize;
  const overallPercent = totalOriginalSize > 0 ? (((totalSaved) / totalOriginalSize) * 100).toFixed(2) : 0;

  console.log('\n📈 Overall Results:');
  console.log(`   - Original Total Size : ${formatBytes(totalOriginalSize)}`);
  console.log(`   - Optimized Total Size: ${formatBytes(totalOptimizedSize)}`);
  console.log(`   - Total Space Saved   : ${formatBytes(totalSaved)} (${overallPercent}% reduction)`);
  console.log(`   - Code Files Updated  : ${updatedCodeFilesCount}`);
  console.log('======================================================\n');
}

run().catch((err) => console.error('Fatal error:', err));

File: scripts/optimize_videos.py
import os
import sys
import subprocess
import shutil

IGNORE_DIRS = {
    'node_modules', '.git', '.next', 'dist', 'build', 'out', '.vscode', '.gemini', 'brain'
}

VIDEO_GIF_EXTENSIONS = {'.mp4', '.mov', '.avi', '.mkv', '.webm', '.m4v', '.gif'}
CODE_EXTENSIONS = {'.html', '.css', '.js', '.jsx', '.ts', '.tsx', '.json', '.md'}

def format_bytes(bytes_val):
    if bytes_val == 0:
        return '0 B'
    sizes = ['B', 'KB', 'MB', 'GB']
    i = 0
    val = float(bytes_val)
    while val >= 1024 and i < len(sizes) - 1:
        val /= 1024.0
        i += 1
    return f"{val:.2f} {sizes[i]}"

def get_ffmpeg_path():
    ffmpeg = shutil.which('ffmpeg')
    if not ffmpeg:
        print("❌ ERROR: FFmpeg executable not found in PATH.")
        print("Please install FFmpeg or install ffmpeg-python / ffmpeg-static.")
        sys.exit(1)
    return ffmpeg

def compress_media(ffmpeg_path, input_path, output_path, is_gif=False, format_type='mp4'):
    if is_gif:
        if format_type == 'webm':
            cmd = [
                ffmpeg_path, '-y', '-i', input_path,
                '-c:v', 'libvpx-vp9', '-b:v', '0', '-crf', '32',
                '-pix_fmt', 'yuva420p', '-an', output_path
            ]
        else:
            cmd = [
                ffmpeg_path, '-y', '-i', input_path,
                '-movflags', 'faststart', '-pix_fmt', 'yuv420p',
                '-vf', 'scale=trunc(iw/2)*2:trunc(ih/2)*2',
                '-c:v', 'libx264', '-crf', '24', '-preset', 'medium', '-an', output_path
            ]
    else:
        if format_type == 'webm':
            cmd = [
                ffmpeg_path, '-y', '-i', input_path,
                '-c:v', 'libvpx-vp9', '-crf', '30', '-b:v', '0',
                '-c:a', 'libopus', '-b:a', '128k', output_path
            ]
        else:
            cmd = [
                ffmpeg_path, '-y', '-i', input_path,
                '-c:v', 'libx264', '-crf', '23', '-preset', 'medium',
                '-movflags', '+faststart', '-c:a', 'aac', '-b:a', '128k',
                '-vf', 'scale=trunc(iw/2)*2:trunc(ih/2)*2', output_path
            ]

    subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True)

def discover_files(root_dir):
    media_files = []
    code_files = []

    for dirpath, dirnames, filenames in os.walk(root_dir):
        dirnames[:] = [d for d in dirnames if d not in IGNORE_DIRS]

        for filename in filenames:
            ext = os.path.splitext(filename)[1].lower()
            full_path = os.path.join(dirpath, filename)
            if ext in VIDEO_GIF_EXTENSIONS:
                media_files.append(full_path)
            elif ext in CODE_EXTENSIONS:
                code_files.append(full_path)

    return media_files, code_files

def main():
    root_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
    ffmpeg_path = get_ffmpeg_path()

    print(f"🎥 Starting Video & GIF Optimization Automation...\n")
    print(f"Using FFmpeg at: {ffmpeg_path}\n")

    media_files, code_files = discover_files(root_dir)

    if not media_files:
        print("✨ No video or GIF files found.")
        return

    print(f"📹 Found {len(media_files)} media file(s) to optimize.")
    print(f"📝 Found {len(code_files)} code file(s) to check for references.\n")

    report = []
    conversion_map = {}
    total_orig = 0
    total_opt = 0

    for old_path in media_files:
        dir_name, file_name = os.path.split(old_path)
        base_name, ext = os.path.splitext(file_name)
        is_gif = ext.lower() == '.gif'
        target_ext = '.mp4' if is_gif else ext.lower()

        temp_output = os.path.join(dir_name, f"{base_name}_optimized{target_ext}")
        final_output = os.path.join(dir_name, f"{base_name}{target_ext}")

        rel_old = os.path.relpath(old_path, root_dir)
        print(f"Optimizing: {rel_old}...")

        try:
            orig_size = os.path.getsize(old_path)
            total_orig += orig_size

            compress_media(ffmpeg_path, old_path, temp_output, is_gif, 'webm' if target_ext == '.webm' else 'mp4')

            opt_size = os.path.getsize(temp_output)
            total_opt += opt_size

            if old_path != final_output:
                shutil.move(temp_output, final_output)
            else:
                shutil.move(temp_output, old_path)

            reduction = ((orig_size - opt_size) / orig_size) * 100
            conversion_map[file_name] = os.path.basename(final_output)

            report.append({
                'orig_name': file_name,
                'opt_name': os.path.basename(final_output),
                'orig_size': orig_size,
                'opt_size': opt_size,
                'reduction': f"{reduction:.2f}%",
                'status': 'SUCCESS'
            })
            print(f"  └─ ✅ Success: {format_bytes(orig_size)} ➔ {format_bytes(opt_size)} ({reduction:.2f}% reduction)")

        except Exception as e:
            print(f"  └─ ❌ Failed: {e}")
            if os.path.exists(temp_output):
                os.remove(temp_output)

    # Update Code
    print("\n🔄 Updating media references in source code files...")
    updated_files = 0
    for code_path in code_files:
        try:
            with open(code_path, 'r', encoding='utf-8', errors='ignore') as f:
                content = f.read()

            modified = False
            for old_name, new_name in conversion_map.items():
                if old_name != new_name and old_name in content:
                    content = content.replace(old_name, new_name)
                    modified = True

            if modified:
                with open(code_path, 'w', encoding='utf-8') as f:
                    f.write(content)
                updated_files += 1
                print(f"  └─ Updated: {os.path.relpath(code_path, root_dir)}")
        except Exception as e:
            print(f"  └─ ❌ Could not update {code_path}: {e}")

    # Summary
    print("\n======================================================")
    print("📊 MEDIA OPTIMIZATION SUMMARY REPORT")
    print("======================================================")
    print(f"{'Original File':<30} {'Optimized File':<30} {'Orig Size':<12} {'Opt Size':<12} {'Reduction'}")
    print("-" * 95)
    for r in report:
        print(f"{r['orig_name']:<30} {r['opt_name']:<30} {format_bytes(r['orig_size']):<12} {format_bytes(r['opt_size']):<12} {r['reduction']}")

    total_saved = total_orig - total_opt
    pct = ((total_saved) / total_orig * 100) if total_orig > 0 else 0
    print("\n📈 Overall Results:")
    print(f"   - Original Total Size : {format_bytes(total_orig)}")
    print(f"   - Optimized Total Size: {format_bytes(total_opt)}")
    print(f"   - Total Space Saved   : {format_bytes(total_saved)} ({pct:.2f}% reduction)")
    print(f"   - Code Files Updated  : {updated_files}")
    print("======================================================\n")

if __name__ == '__main__':
    main()
